File: klipian/roundtrip.py
"""Manual round-trip through Claude, without an API key.

Workflow:

    klipian brief video.mp4      ->  out/<video>/brief-claude.md
                                     drop into Claude, ask it to work
    (Claude replies with JSON)
    klipian import video.mp4 reply.json
                                 ->  out/<video>/kandidat.json

Why this matters beyond just avoiding costs: you see Claude's reasoning
before anything is rendered, and you can argue with it. API-based tools
hide that step.

Key division of labor:

    Claude   gives APPROXIMATE times in min:sec
    klipian  snaps those times to the nearest word boundary

Claude doesn't need millisecond precision -- that's not its job, and forcing
it only makes the results brittle. The per-word timestamps we already have in
cache handle the final rounding.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

from .models import Transcript, Word


def fmt_time(seconds: float) -> str:
    t = max(0, int(round(seconds)))  # clamp to 0 to avoid negative values
    return f"{t // 60}:{t % 60:02d}" if t < 3600 else f"{t // 3600}:{(t % 3600) // 60:02d}:{t % 60:02d}"


def seconds(text: str) -> float:
    """Accept 1:07, 01:07, or 1:24:10."""
    parts = [float(x) for x in str(text).strip().split(":")]
    result = 0.0
    for b in parts:
        result = result * 60 + b
    return result


# --------------------------------------------------------------------------
# EXPORT: files dropped into Claude
# --------------------------------------------------------------------------

def build_brief(transcript: Transcript, rubric: Path, video: Path) -> str:
    """A single self-contained file: task, rubric, transcript, and expected
    answer format. The user just drops it in and says "do it" -- no
    additional explanation needed."""

    rubric_text = rubric.read_text(encoding="utf-8") if rubric.exists() else ""
    # strip rubric title so it doesn't clash with the brief title
    rubric_text = re.sub(r"\A# .*?\n+", "", rubric_text)
    # Strip rubric intro paragraph that addresses the USER, not Claude.
    # "Edit this file if the results don't match your taste" is meaningless
    # to whoever is reading this brief.
    rubric_text = re.sub(r"\AIni yang dibaca Claude.*?---\s*\n+", "", rubric_text, flags=re.S)

    lines = []
    for seg in transcript.segments:
        text = seg.text.strip()
        if text:
            lines.append(f"[{fmt_time(seg.start)}] {text}")
    transcript_lines = "\n".join(lines)

    return f"""# Find clips — {video.name}

Hello. Please read the transcript at the bottom of this file and pick moments
worth turning into short vertical videos.

**Source duration:** {fmt_time(transcript.duration)} · **{len(transcript.words)} words**

The times you give don't need to be precise. Approximate min:sec is fine;
klipian will snap the cut points to the nearest word boundary.

---

## Evaluation rubric

{rubric_text}

---

## Answer format

Reply with **a single JSON block** exactly like this, no extra explanation
outside the block. Save as `.json`, then import back into klipian.

```json
{{
  "source": "{video.name}",
  "clips": [
    {{
      "start": "0:12",
      "end": "1:07",
      "title": "Lose 300M Because of Timing",
      "hook": "exact quote from transcript",
      "scores": {{ "hook": 9, "complete": 8, "payoff": 9, "emotion": 8, "duration": 9 }},
      "reason": "one or two sentences for a human to read"
    }}
  ]
}}
```

---

## Transcript

{transcript_lines}
"""


# --------------------------------------------------------------------------
# IMPORT: reading Claude's reply
# --------------------------------------------------------------------------

@dataclass
class Candidate:
    start: float
    end: float
    title: str
    hook: str
    scores: dict
    reason: str

    @property
    def duration(self) -> float:
        return self.end - self.start

    @property
    def total(self) -> float:
        n = []
        for v in self.scores.values():
            try:                       # "9" is counted, "nine" is skipped
                n.append(float(v))
            except (TypeError, ValueError):
                pass
        return round(sum(n) / len(n), 1) if n else 0.0


def extract_json(text: str) -> dict:
    """Claude's replies are often wrapped in code fences or prefixed with
    prose. Grab the largest JSON object found, don't require a clean file."""
    # Look inside code fences first. Use a lazy match ending at the closing
    # fence -- identical to extractJSON() in ui/roundtrip.js. The old greedy
    # version spanned from the first fence to the LAST fence, so replies
    # with two code blocks failed in the CLI but worked in the UI.
    fence = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", text, re.S)
    if fence:
        return json.loads(fence.group(1))
    # Fallback: find the largest { ... } (greedy rfind)
    first, end = text.find("{"), text.rfind("}")
    if first == -1 or end <= first:
        raise ValueError("No JSON block found in that file.")
    return json.loads(text[first:end + 1])


# How far a cut point may be shifted to snap to a word boundary. A word
# boundary farther than this means Claude pointed at silence, not at a word
# that was slightly off -- the original number is correct there. Without
# this limit, a 5-second clip in the middle of a long pause once ballooned
# to 79 seconds.
SNAP_MAX = 2.0


def snap_to_word(words: list[Word], moment: float, side: str) -> float:
    """Snap to the nearest word boundary, at most SNAP_MAX seconds away.

    side="start" -> to the START of the word being/will-be spoken
    side="end"   -> to the END of the word that just finished

    This is the part Claude can't do: it doesn't have per-word timestamps.
    We do.
    """
    if not words:
        return moment
    if side == "start":
        candidates = [w.start for w in words]
    else:
        candidates = [w.end for w in words]
    nearest = min(candidates, key=lambda t: abs(t - moment))
    return nearest if abs(nearest - moment) <= SNAP_MAX else moment


def parse_reply(transcript: Transcript, text: str) -> list[Candidate]:
    """English keys are primary; Indonesian keys are still accepted as
    fallback so older Claude replies can still be imported."""
    data = extract_json(text)
    raw = data.get("clips") or data.get("klip") or []
    if not raw:
        raise ValueError('The JSON has no "clips" list.')

    def pick(d, *keys):
        # Take the FIRST key that EXISTS, not the first truthy one: start=0
        # (a number) is valid but falsy, `a or b` would wrongly skip it.
        for key in keys:
            if key in d:
                return d[key]
        return None

    words = transcript.words
    result: list[Candidate] = []
    for k in raw:
        try:
            m = seconds(pick(k, "start", "mulai"))
            s = seconds(pick(k, "end", "selesai"))
        except (TypeError, ValueError):
            continue
        if s <= m:
            continue

        # Checked AGAIN after snapping: the two sides move independently,
        # so a range that was valid may become inverted. If that happens,
        # use Claude's original numbers -- a slight miss is better than a
        # broken candidate that only fails at render time.
        a = round(snap_to_word(words, m, "start"), 3)
        b = round(snap_to_word(words, s, "end"), 3)
        if b <= a:
            a, b = m, s

        result.append(Candidate(
            start=a,
            end=b,
            title=(k.get("title") or k.get("judul") or "Untitled").strip(),
            hook=(k.get("hook") or "").strip(),
            # Explicit check: empty dict {} is valid but falsy, don't use or
            scores=k["scores"] if "scores" in k else (k.get("skor") or {}),
            reason=(k.get("reason") or k.get("alasan") or "").strip(),
        ))

    result.sort(key=lambda x: x.total, reverse=True)
    return result


def save_candidates(candidates: list[Candidate], video: Path, dest: Path) -> Path:
    """The format is intentionally identical to the UI prototype, so it can
    be read directly without translation."""
    dest.parent.mkdir(parents=True, exist_ok=True)
    payload = json.dumps({
        "source": video.name,
        "candidates": [{
            "title": k.title,
            "hook": k.hook,
            "in": fmt_time(k.start),
            "out": fmt_time(k.end),
            "start_sec": k.start,
            "end_sec": k.end,
            "dur": round(k.duration),
            "total": k.total,
            "scores": k.scores,
            "reason": k.reason,
        } for k in candidates],
    }, ensure_ascii=False, indent=1)
    # Write to tmp then rename (atomic) -- a failed import won't leave a
    # truncated candidates.json. Same pattern as Transcript.save.
    tmp = dest.with_suffix(".tmp")
    tmp.write_text(payload, encoding="utf-8")
    tmp.replace(dest)
    return dest

File: ui/js/editor/roundtrip.js
/* klipian — round-trip via Claude web
   ==========================================================================
   Flow:

     1  drop video on the Prepare screen
     2  app builds a file containing rubric + transcript            -> Download
     3  that file is dropped into Claude web, ask it to "work on it"
     4  Claude web replies with JSON
     5  JSON is pasted here                                        -> Import
     6  app cuts according to JSON, clips appear in Candidates

   No API key in this path. The only cost is your existing Claude subscription.

   The transcript is REAL: read from workspace/cache/ populated by `klipian transcribe`.
   The server serves the project root, so the UI can reach it.
   ========================================================================== */

const ROOT = "../";                       // ui/ -> project root
let realTranscript = null;                // { duration, segments[], words[] }

/* ---------- finding a transcript that matches the dropped file ----
   The cache filename contains a fingerprint, so we match via its stem:
   radityadika-podcast.mp4 -> radityadika-podcast.*.transcript.json */

async function listCache() {
  // The klipian server provides the listing via API. If the UI is opened via
  // a plain http.server, fall back to its HTML directory listing.
  try {
    const d = await (await fetch("/api/cache")).json();
    if (d.transcript?.length) return d.transcript;
  } catch { /* fall back to directory listing */ }
  try {
    const html = await (await fetch(ROOT + "workspace/cache/")).text();
    return [...html.matchAll(/href="([^"]+\.transcript\.json)"/g)].map((m) => m[1]);
  } catch {
    return [];
  }
}

async function findTranscript(videoName) {
  const stem = videoName.replace(/\.[^.]+$/, "");
  const all = await listCache();
  const matched = all.find((f) => decodeURIComponent(f).startsWith(stem + "."));
  if (!matched) return null;
  const d = await (await fetch(ROOT + "workspace/cache/" + matched)).json();
  d.words = d.segments.flatMap((s) => s.words || []);
  return d;
}

/* ---------- step 2: building the file for Claude ---------- */

const fmtStamp = (d) => {
  const t = Math.round(d);
  return t < 3600
    ? `${Math.floor(t / 60)}:${String(t % 60).padStart(2, "0")}`
    : `${Math.floor(t / 3600)}:${String(Math.floor((t % 3600) / 60)).padStart(2, "0")}:${String(t % 60).padStart(2, "0")}`;
};

async function buildBrief(name) {
  // Just one rubric now that categories were removed. This one works from the transcript,
  // so it applies to any source that has audio.
  const rubricFile = "dialog-podcast.md";
  let rubric = "";
  try { rubric = await (await fetch(ROOT + "prompts/rubrik/" + rubricFile)).text(); } catch {}
  rubric = rubric.replace(/^# .*\n+/, "").replace(/^Ini yang dibaca Claude[\s\S]*?---\s*\n+/, "");

  const row = realTranscript.segments
    .filter((s) => s.text.trim())
    .map((s) => `[${fmtStamp(s.start)}] ${s.text.trim()}`)
    .join("\n");

  // SEPARATE signal from the transcript -- purely from audio volume (see
  // klipian/audio_energy.py), triggered automatically once per video alongside
  // transcription. Whisper rarely captures laughter/strong reactions as reliable
  // text; this is actual measured evidence, not punctuation guesswork.
  // Empty if not yet analyzed OR if there simply are no prominent moments --
  // both are the same here: the section is skipped.
  let energyBlock = "";
  try {
    const d = await (await fetch(`/api/audio-energy?video=${encodeURIComponent(name)}`)).json();
    if (d.moments?.length) {
      const energyList = d.moments
        .map((m) => `- [${fmtStamp(m.start)}] – [${fmtStamp(m.end)}]`).join("\n");
      energyBlock = `

---

## Prominent audio-energy moments (from volume analysis, NOT the transcript)

The following time ranges have sound energy well above this video's own
average -- could be audience laughter, cheering, a loud reaction, or a
dramatic moment. This is a SUPPORTING signal, not a certain fact: cross-
check it against the transcript's sentences around that time before using
it as a reason for a score.

${energyList}
`;
    }
  } catch { /* no backend means nothing to fetch -- brief still works without this section */ }

  return `# Find clips — ${name}

Hello. Please read the transcript at the bottom of this file and pick moments
worth turning into short vertical videos.

**Source duration:** ${fmtStamp(realTranscript.duration)} · **${realTranscript.words.length} words**

The times you give don't need to be precise. Approximate min:sec is fine;
klipian will snap the cut points to the nearest word boundary.

---

## Evaluation rubric

${rubric}

---

## Answer format

Reply with **a single JSON block** exactly like this, no extra explanation
outside the block.

\`\`\`json
{
  "source": "${name}",
  "clips": [
    {
      "start": "0:12",
      "end": "1:07",
      "title": "Lose 300M Because of Timing",
      "hook": "exact quote from transcript",
      "scores": { "hook": 9, "complete": 8, "payoff": 9, "emotion": 8, "duration": 9 },
      "reason": "one or two sentences for a human to read"
    }
  ]
}
\`\`\`
${energyBlock}
---

## Transcript

${row}
`;
}

/* ---------- step 5: reading Claude's reply ---------- */

const toSeconds = (t) => {
  // Empty string/space -> NaN, not 0. Number("") is 0, so without this guard
  // "start": "" would pass Number.isFinite and become a clip starting at 0:00.
  const s = String(t ?? "").trim();
  if (!s) return NaN;
  return s.split(":").reduce((a, b) => a * 60 + Number(b), 0);
};

function extractJSON(text) {
  // JSON.parse throws a technical English error message ("Expected property
  // name or '}' ... at position 14"). Wrapped so the user sees an
  // actionable sentence instead.
  const parse = (s) => {
    try {
      return JSON.parse(s);
    } catch {
      throw new Error("The JSON block is there, but its content is broken — copy " +
                      "Claude's reply again from the start to the end of the block.");
    }
  };
  const fence = text.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
  if (fence) return parse(fence[1]);
  const a = text.indexOf("{"), b = text.lastIndexOf("}");
  if (a === -1 || b <= a) throw new Error("No JSON block found in that text.");
  return parse(text.slice(a, b + 1));
}

/* Snap to nearest word boundary. This is what Claude cannot do:
   it has no per-word timestamps. We do. */
/* How far a cut point may be shifted. Must match SNAP_MAX in
   klipian/roundtrip.py -- if one is changed, change both. A word boundary
   farther away than this means Claude pointed at silence, not at a word
   that is only slightly off. */
const SNAP_MAX = 2.0;

function snapToWord(time, side) {
  const w = realTranscript?.words;
  if (!w?.length || !Number.isFinite(time)) return time;
  const candidate = side === "start" ? w.map((x) => x.start) : w.map((x) => x.end);
  const nearest = candidate.reduce((a, b) =>
    (Math.abs(b - time) < Math.abs(a - time) ? b : a));
  return Math.abs(nearest - time) <= SNAP_MAX ? nearest : time;
}

/* Indonesian keys (klip, mulai, judul, ...) are intentionally still accepted as
   fallback. The prompt now asks for English keys, but Claude replies that were
   saved before the rename must still be importable -- and Claude occasionally
   answers using terms from the prompt prose. */
function importJSON(text) {
  const data = extractJSON(text);
  const raw = data.clips || data.klip || [];
  if (!raw.length) throw new Error('JSON does not have a "clips" list.');

  const result = raw.map((k) => {
    // Clips without timestamps are skipped, not accepted as 0:00. Without this
    // guard toSeconds(undefined) returns NaN, and the always-false NaN comparison
    // causes snapToWord to return the first word -- one field Claude forgot to
    // write turns into a clip starting at the beginning of the video.
    const m0 = toSeconds(k.start ?? k.mulai);
    const s0 = toSeconds(k.end ?? k.selesai);
    if (!Number.isFinite(m0) || !Number.isFinite(s0) || s0 <= m0) return null;

    // Re-check after snapping: each side moves independently, so a range that
    // was valid before can end up inverted. If so, fall back to the original
    // numbers from Claude.
    let m = snapToWord(m0, "start");
    let s = snapToWord(s0, "end");
    if (s <= m) { m = m0; s = s0; }

    // Scores are COERCED to numbers here, at the boundary where external data
    // enters. Claude occasionally writes "nine" or "9" (string), and non-numeric
    // values slip through until renderBoard calls value.toFixed() and throws --
    // the Candidates board is permanently broken until the page reloads.
    const scores = k.scores || k.skor || {};
    const num = (v) => { const n = Number(v); return Number.isFinite(n) ? n : 0; };
    const nums = Object.values(scores)
      .map(Number).filter(Number.isFinite);
    return {
      title: (k.title || k.judul || "Untitled").trim(),
      hook: (k.hook || "").trim(),
      in: fmtStamp(m), out: fmtStamp(s),
      startSec: m, endSec: s,
      dur: Math.round(s - m),
      total: nums.length ? +(nums.reduce((a, b) => a + b, 0) / nums.length).toFixed(1) : 0,
      scores: {
        hook: num(scores.hook),
        complete: num(scores.complete ?? scores.self_contained),
        payoff: num(scores.payoff),
      },
      reason: (k.reason || k.alasan || "").trim(),
      spans: [{ start: m, end: s }],   // one whole piece until split
    };
  // AI order is preserved, NOT re-sorted by score. Scores are not displayed
  // on the Clip screen, so re-sorting only makes "recommendation #1" on
  // screen differ from #1 in the JSON.
  }).filter((k) => k && k.dur > 0);

  if (!result.length) throw new Error("All clips have zero or negative duration.");
  return result;
}

/* ---------- wiring results into the rest of the app ---------- */

function applyCandidates(candidates) {
  const d = DATA;
  d.candidates = candidates;
  // realTranscript can be null if JSON is imported before the transcript loads;
  // importJSON/snapToWord tolerate it, so don't blindly deref here either.
  const dur = realTranscript?.duration;
  d.marks = candidates.map((k) => ({
    pos: dur ? (k.startSec / dur) * 100 : 0,
    scores: k.total,
    label: (k.title || "").split(" ").slice(0, 3).join(" "),
  }));

  // Cut screen transcript uses real words around the top clip


  renderList(); renderPreview();
  if (typeof setClip === "function") setClip(candidates[0]);

}


/* ---------- wiring controls ---------- */

async function prepareExport(videoName) {
  const panel = $("#exportPanel"), button = $("#downloadBrief"), note = $("#exportNote");
  note.textContent = "searching for transcript ...";
  const previous = realTranscript;
  realTranscript = await findTranscript(videoName);

  if (!realTranscript) {
    panel.dataset.ready = "false";
    button.disabled = true;
    note.textContent = `No transcript yet. Run this first: klipian transcribe ${videoName}`;
    return;
  }
  panel.dataset.ready = "true";
  button.disabled = false;
  note.textContent =
    `${realTranscript.words.length.toLocaleString("en")} words · ${fmtStamp(realTranscript.duration)} · drop the JSON to your AI to analyze it`;
  $("#importPanel").dataset.ready = "true";

  // If the user already added manual clips BEFORE this transcript arrived
  // (e.g. transcription was still running when "Create manual clip" was pressed),
  // the captions are empty because kataResult() has nothing to map yet -- and
  // nothing triggers a redraw once the transcript finally becomes available.
  // An explicit redraw here closes that gap.
  if (!previous && typeof RESULT !== "undefined" && RESULT.length) {
    if (typeof drawCaption === "function") drawCaption();
    if (typeof renderCaptions === "function") renderCaptions();
  }
}

$("#downloadBrief").addEventListener("click", async () => {
  const name = chosenSource?.name || "video.mp4";
  const text = await buildBrief(name);
  const a = document.createElement("a");
  a.href = URL.createObjectURL(new Blob([text], { type: "text/markdown" }));
  a.download = `brief-claude-${name.replace(/\.[^.]+$/, "")}.md`;
  a.click();
  URL.revokeObjectURL(a.href);
  $("#exportNote").textContent =
    `File downloaded · ${(text.length / 1024).toFixed(0)} KB · ~${Math.round(text.length / 3.5).toLocaleString("en")} tokens`;
});

$("#pasteJSON").addEventListener("input", (e) => {
  $("#importBtn").disabled = !e.target.value.trim();
  $("#importNote").dataset.error = "false";
  $("#importNote").textContent = "";
});

$("#importBtn").addEventListener("click", () => {
  const note = $("#importNote");
  try {
    const candidates = importJSON($("#pasteJSON").value);
    applyCandidates(candidates);
    note.dataset.error = "false";
    note.textContent = `${candidates.length} suggestions imported`;
    if (typeof renderRecommendations === "function") renderRecommendations();
    toScreen("klip");          // recommendations are not the final destination, result is
  } catch (err) {
    note.dataset.error = "true";
    note.textContent = err.message;
  }
});

/* .json files can be dropped directly onto the paste box */
$("#pasteJSON").addEventListener("drop", async (e) => {
  const f = e.dataTransfer?.files?.[0];
  if (!f) return;
  e.preventDefault();
  e.stopPropagation();
  $("#pasteJSON").value = await f.text();
  $("#importBtn").disabled = false;
});
